File: ios/TheSlowWire/TheSlowWire/Views/ArticleRow.swift
import SwiftUI

// One entry in the day's journal: folio numeral, headline, the product's own
// why-line, and one apparatus meta line. Every entry keeps one type size;
// rank shows as position and as the lead's full-gold numeral, never as scale.
struct ArticleRow: View {
    let story: Story
    let number: Int
    let isLead: Bool

    private var article: FeedArticle { story.lead }

    var body: some View {
        HStack(alignment: .top, spacing: 14) {
            Text("\(number)")
                .font(.displayItalicish(isLead ? 30 : 20))
                .italic()
                .foregroundStyle(isLead ? Color.gold : Color.goldSoft.opacity(0.55))
                .frame(width: 34, alignment: .trailing)
                .padding(.top, isLead ? 0 : 2)
                .accessibilityHidden(true)

            VStack(alignment: .leading, spacing: 6) {
                Text(article.title)
                    .font(.display(17))
                    .foregroundStyle(Color.ink)
                    .fixedSize(horizontal: false, vertical: true)

                if let why = article.whyItMatters, !why.isEmpty {
                    Text(why)
                        .font(.prose(13.5))
                        .italic()
                        .foregroundStyle(Color.inkSoft)
                        .lineLimit(2)
                        .fixedSize(horizontal: false, vertical: true)
                }

                metaLine
            }

            Spacer(minLength: 0)
        }
        .padding(.vertical, 13)
        .contentShape(Rectangle())
    }

    @ViewBuilder private var metaLine: some View {
        let points = article.points
        HStack(spacing: 6) {
            if !article.sourceName.isEmpty {
                Apparatus(article.sourceName, size: 10.5, medium: true)
                    .foregroundStyle(Color.ink)
            }
            if let points {
                Apparatus("· \(points) points", size: 10.5)
                    .foregroundStyle(Color.muted)
            }
            if story.otherSources > 0 {
                Apparatus("· +\(story.otherSources) more", size: 10.5)
                    .foregroundStyle(Color.muted)
            }
        }
        .padding(.top, 2)
    }
}

#Preview {
    let lead = FeedArticle(
        urlHash: "abc", url: "https://example.com",
        title: "The day's lead story under its gold announcement",
        summary: "", imageUrl: nil, source: "hn", sourceName: "Hacker News",
        category: "community", section: "ai", publishedAt: nil,
        clusterId: "2026-08-28#lead", corroborationToday: 3,
        whyItMatters: "One sentence on why this story matters today.",
        score: 90, scoreVersion: "v1", points: 703, pointsImputed: false,
        llmImportance: 70, firstSeenAt: "2026-08-28T12:00:00.000Z"
    )
    let second = FeedArticle(
        urlHash: "def", url: "https://example.com",
        title: "An ordinary entry stays at the same size",
        summary: "", imageUrl: nil, source: "hn", sourceName: "Hacker News",
        category: "community", section: "ai", publishedAt: nil,
        clusterId: "2026-08-28#second", corroborationToday: nil,
        whyItMatters: "Position, not scale, is the rank.",
        score: 60, scoreVersion: "v1", points: 240, pointsImputed: false,
        llmImportance: 55, firstSeenAt: "2026-08-28T12:00:00.000Z"
    )
    return VStack(spacing: 0) {
        ArticleRow(story: Story(lead: lead, others: [second], rank: 1), number: 1, isLead: true)
        Rectangle().fill(Color.hairSoft).frame(height: 0.5)
        ArticleRow(story: Story(lead: second, others: [], rank: 2), number: 2, isLead: false)
    }
    .padding(.horizontal, 20)
    .background(Color.ground)
}

File: android/TheSlowWire/app/src/main/java/com/eienmosu/theslowwire/ui/section/ArticleRow.kt
package com.eienmosu.theslowwire.ui.section

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import com.eienmosu.theslowwire.model.Story
import com.eienmosu.theslowwire.ui.theme.Palette
import com.eienmosu.theslowwire.ui.theme.apparatus
import com.eienmosu.theslowwire.ui.theme.current
import com.eienmosu.theslowwire.ui.theme.display
import com.eienmosu.theslowwire.ui.theme.folio
import com.eienmosu.theslowwire.ui.theme.prose

/**
 * One entry in the day's journal — not a card, a row on the page: the folio
 * numeral, the headline, the product's own why-line, and one apparatus meta
 * line. The scraped summary and the thumbnail deliberately stay out of the
 * feed rows (they live in the reading room), matching the web and iOS exactly.
 *
 * The lead is announced by ground, never by scale: its numeral is full gold
 * where the others are soft, and every headline keeps one type size.
 */
@Composable
fun ArticleRow(
    story: Story,
    isLead: Boolean,
    modifier: Modifier = Modifier,
) {
    val palette = Palette.current
    val article = story.lead

    Row(modifier.padding(vertical = 16.dp)) {
        Text(
            text = story.rank.toString(),
            style = folio(if (isLead) 38 else 24),
            color = if (isLead) palette.gold else palette.goldSoft,
            textAlign = TextAlign.End,
            modifier = Modifier.width(44.dp),
        )
        Spacer(Modifier.width(14.dp))
        Column(Modifier.fillMaxWidth()) {
            Text(
                text = article.title,
                style = display(17, FontWeight.Bold),
                color = palette.ink,
            )
            article.whyItMatters?.takeIf { it.isNotBlank() }?.let { why ->
                Spacer(Modifier.height(6.dp))
                Text(text = why, style = prose(14), color = palette.inkSoft)
            }
            Spacer(Modifier.height(8.dp))
            MetaLine(story)
        }
    }
}

/**
 * source · N points · +K more. The source name is the bold part wherever it
 * sits, never "whichever part comes first" — a positional rule that was a real
 * bug on the web until an agent caught it.
 */
@Composable
private fun MetaLine(story: Story) {
    val palette = Palette.current
    val article = story.lead

    val text = buildAnnotatedString {
        withStyle(SpanStyle(color = palette.ink, fontWeight = FontWeight.Medium)) {
            append(article.sourceName.uppercase())
        }
        article.points?.let { append(" · ${it} POINTS") }
        if (story.otherSources > 0) append(" · +${story.otherSources} MORE")
    }
    Text(text = text, style = apparatus(11), color = palette.muted)
}
